src/launcher/plugin.ts

```diff
@@ -0,0 +1,298 @@
+  });
+
+  let launcherWidget = new LauncherWidget()
+
+  launcherWidget.model = launcherModel;
+  launcherWidget.id = 'landing-jupyterlab-widget';
+  launcherWidget.title.label = 'Launcher';
+
+  // Hardcoded defaults.
+  let names = [
+    'Notebook',
+    'Console',
+    'Terminal',
+    'Text Editor',
+  ];
+
+  let actions = [
+    'file-operations:new-notebook',
+    `console:create-${services.kernelspecs.default}`,
+    'terminal:create-new',
+    'file-operations:new-text-file',
+  ]
+
+  app.commands.addCommand('jupyterlab-launcher:add-item', {
+    label: 'Add Launcher Item',
+    execute: (args) => {
+      launcherModel.add(args['name'] as string, args['action'] as string,
+                        args['args'] as JSONObject, args['imgName'] as string);
+    }
+  });
+
+  for (let i in names) {
+    // Note: we do not retain a handle on the items added by default, which
+    // means we have to way of removing them after the fact.
+    app.commands.execute('jupyterlab-launcher:add-item', {name: names[i], action: actions[i]});
+  }
+
+
+  app.commands.addCommand('jupyterlab-launcher:show', {
+    label: 'Show Launcher',
+    execute: () => {
+      if (!launcherWidget.isAttached) {
+        app.shell.addToLeftArea(launcherWidget);
+      }
+      app.shell.activateLeft(launcherWidget.id);
+    }
+  });
+
+  palette.addItem({
+    command: 'jupyterlab-launcher:show',
+    category: 'Help'
+  });
+
+  app.shell.addToLeftArea(launcherWidget);
+  return launcherModel;
+}
+
+
+/**
+ * Simple encapsulation of name and callback of launcher entries.
+ */
+class LauncherItem {
+
+  readonly name: string;
+ 
+  readonly clickCallback: () => void;
+  
+  readonly imgName: string;
+ 
+  constructor(name: string, clickCallback: () => void, imgName?: string) {
+    this.name = name;
+    this.clickCallback = clickCallback;
+    if (imgName) {
+      this.imgName = imgName;
+    } else {
+      this.imgName = 'jp-Image' + name.replace(/\ /g, '');
+    }
+  }
+}
+
+
+/**
+ * LauncherModel keeps track of the path to working directory and has a list of
+ * LauncherItems.
+ */
+class LauncherModel extends VDomModel implements ILauncher {
+  items: LauncherItem[] = [];
+  path: string = 'home';
+  app: JupyterLab;
+  /**
+   * Convenience method to add a launcher with a given name and callback.
+   *
+   * This keeps plugins who wish to register themselves in the Launcher from
+   * having to import both LauncherItems and LauncherItem.
+   */
+  add(name: string, action: string, args?: JSONObject, imgName?: string) : IDisposable {
+    let clickCallback = () => { this.app.commands.execute( action, args) };
+    return this.addItem(new LauncherItem(name, clickCallback, imgName));
+  }
+
+  /**
+   * Add a new launcher and trigger re-render event for parent widget.
+   *
+   * @returns A Disposable which can be called to remove this new item from
+   * the Launcher and trigger another re-render event.
+   */
+  addItem(item: LauncherItem) : IDisposable {
```

Comment:
Do you think we need separate `addItem` and `add` methods? Is the only difference that one is command based, the other callback based? Would it make sense to simplify and offer a single one that simple takes the `LauncherItem` (make it command based) and automatically hooks up the callback for the command?